alacritty_terminal/src/selection.rs

```diff
@@ -81,89 +85,175 @@ pub trait Dimensions {
+        Selection::Normal {
             region: Range {
-                start: Anchor::new(location.into(), side),
-                end: Anchor::new(location.into(), side),
+                start: Anchor::new(location.into(), SelectionType::Simple, side),
+                end: Anchor::new(location.into(), SelectionType::Simple, side),
             },
         }
     }
 
     pub fn block(location: Point<usize>, side: Side) -> Selection {
         Selection::Block {
             region: Range {
-                start: Anchor::new(location.into(), side),
-                end: Anchor::new(location.into(), side),
+                start: Anchor::new(location.into(), SelectionType::Simple, side),
+                end: Anchor::new(location.into(), SelectionType::Simple, side),
             },
         }
     }
 
     pub fn semantic(point: Point<usize>) -> Selection {
-        Selection::Semantic { region: Range { start: point.into(), end: point.into() } }
+        Selection::Normal {
+            region: Range {
+                start: Anchor::new(point.into(), SelectionType::Semantic, Side::Right),
+                end: Anchor::new(point.into(), SelectionType::Semantic, Side::Right),
+            },
+        }
     }
 
     pub fn lines(point: Point<usize>) -> Selection {
-        Selection::Lines { region: Range { start: point.into(), end: point.into() } }
+        Selection::Normal {
+            region: Range {
+                start: Anchor::new(point.into(), SelectionType::Lines, Side::Right),
+                end: Anchor::new(point.into(), SelectionType::Lines, Side::Right),
+            },
+        }
     }
 
     pub fn update(&mut self, location: Point<usize>, side: Side) {
-        // Always update the `end`; can normalize later during span generation.
         match *self {
-            Selection::Simple { ref mut region } | Selection::Block { ref mut region } => {
-                region.end = Anchor::new(location.into(), side);
-            },
-            Selection::Semantic { ref mut region } | Selection::Lines { ref mut region } => {
-                region.end = location.into();
-            },
+            Selection::Normal {ref mut region } | Selection::Block {ref mut region } => {
+                // Always update the `end`; can normalize later during span generation.
+                // Always set side; can be ignored later if needed.
+                region.end = Anchor::new(location.into(), region.end.ty.clone(), side);
+            }
+        }
+    }
+
+    pub fn update_as(&mut self, location: Point<usize>, side: Side, selection_type: SelectionType) {
+        match *self {
+            Selection::Normal {ref mut region } | Selection::Block { ref mut region } => {
+                region.end.ty = selection_type;
+            }
+        }
+        self.update(location, side);
+    }
+
+    // Expand selection following the similar rules as creating a span
+    pub fn expand_selection(term: &mut Term) {
```

Comment:
Formatting of this function is very messed up. Additionally the control flow is incredibly confusing, so I don't really feel like it makes much sense to review it in its current state. This should be simplified to make it easier to maintain and understand. Extraction in sub functions might be one option.

Reply:
Yeah, I agree that this function is messy. I'll do some cleanup on it.